edit/candidate.go

```diff
@@ -101,39 +101,29 @@ func outputComplexCandidate(ec *eval.EvalCtx,
 	ec.OutputChan() <- c
 }
 
-func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, completer string, pattern string,
-	cands []rawCandidate, q parse.PrimaryType) ([]*candidate, error) {
+func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, matcher eval.CallableValue, pattern string,
+	cands chan eval.Value, q parse.PrimaryType) ([]*candidate, error) {
 
-	matcher, ok := ed.lookupMatcher(completer)
-	if !ok {
-		return nil, errMatcherMustBeFn
-	}
-
-	input := make(chan eval.Value, len(cands))
 	ports := []*eval.Port{
-		{Chan: input}, {File: os.Stdout}, {File: os.Stderr}}
+		{Chan: cands}, {File: os.Stdout}, {File: os.Stderr}}
 	ec := eval.NewTopEvalCtx(ev, "[editor matcher]", "", ports)
 
-	// TODO(xiaq): Revise the API of completers so that the complter write
-	// directly to a channel and run concurrently with the matcher.
-	for _, cand := range cands {
-		input <- eval.String(cand.text())
-	}
-	close(input)
-
 	args := []eval.Value{eval.String(pattern)}
 	values, err := ec.PCaptureOutput(matcher, args, eval.NoOpts)
 	if err != nil {
+		logger.Printf("tw; %s\n", err)
 		return nil, err
-	} else if len(values) != len(cands) {
-		return nil, errIncorrectNumOfResults
 	}
 
 	var filtered []*candidate
-	for i, value := range values {
-		if eval.ToBool(value) {
-			filtered = append(filtered, cands[i].cook(q))
+	for _, value := range values {
```

Comment:
But I think returning the filtering results is simpler and it doesn't add additional complexity to custom matcher writing. It just returns what they want instead of the boolean value of what they want. There is also another benefit: matcher could also return additional candidates according to the original input.

Also there is no way to obtain the original raw candidates here.